tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +429,275 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+                                           const Dimension<3>& input_dims,
+                                           T* output) {
+    assert(false &&
+           "BatchNarrowMatrixTransposeDispatcher has requested an unexpected "
+           "launch configuration. ");
+  }
+};
+
+// A helper function to make RunSwapDimension1And2InTensor3 concise. This
+// helper function looks at the data type and input matrix sizes and decides
+// the thread numbers and tile sizes to use.
+template <typename T>
+void SwapDimension1And2InTensor3WithNarrowMatrices(
+    const GPUDevice& d, const T* input, const Dimension<3>& input_dims,
+    T* output, const int kMinDimensionToUseTiles) {
+  // Define available tile sizes here for each size of data type supported:
+  std::vector<std::pair<int, int>> tile_spec =
+      []() -> std::vector<std::pair<int, int>> {
+    switch (sizeof(T)) {
+      case 1:
+        return {{32, 15}, {64, 15}, {128, 15}, {256, 15}, {512, 4}, {1024, 2}};
+      case 2:
+        return {{32, 15}, {64, 15}, {128, 15}, {256, 15}, {512, 4}, {1024, 2}};
+      case 4:
+        return {{32, 15}, {64, 15}, {128, 15}, {256, 15}, {512, 4}, {1024, 2}};
+      case 8:
+        return {{32, 15}, {64, 15}, {128, 8}, {256, 4}, {512, 2}};
+      case 16:
+        return {{32, 4},  {64, 4},  {128, 4}, {256, 2}};
+    }
+    return {};
+  }();
+
+  int tile_long_side_len = 0;
+  int tile_short_side_len = 0;
+  float lowest_cost = std::numeric_limits<float>::max();
+  int data_long_side = max(input_dims[1], input_dims[2]);
+
+  for (auto tile_size_pair : tile_spec) {
+    int proposed_tile_long_side_len = tile_size_pair.first;
+
+    // Threads that will not be doing anything useful when reading the matrix
+    // because the thread block size is bigger than the data block size.
+    int wasted_threads = (data_long_side -
+                          MathUtil::FloorOfRatio<int>(
+                              data_long_side, proposed_tile_long_side_len) *
+                              proposed_tile_long_side_len);
+
+    int num_full_tiles = MathUtil::FloorOfRatio<int>(
+        data_long_side, proposed_tile_long_side_len);
+
+    float cost = 0;
+
+    // However, if we can execute two or more full tiles, then we gladly
+    // accept any number of wasted thread and ignore its cost.
+    if (num_full_tiles <= 1) cost = wasted_threads;
+
+    // Using less and equal here because given the same cost, we would like to
+    // launch as many threads as possible.
+    if (cost <= lowest_cost) {
+      tile_long_side_len = proposed_tile_long_side_len;
+      tile_short_side_len = tile_size_pair.second;
+      lowest_cost = cost;
+    }
+  }
+
+  // Request tile sizes such that the longer side of threadblock aligns with
+  // the longer side of input data block to maximize read throughput.
+  // The ideal tile shape to request is one with its length of the shorter
+  // side of the tile being equal to the length of the shorter side of the
+  // input matrix.
```

Comment:
Suggest: "The ideal tile shape is one where the length of the shorter side of the tile is equal to the length of the shorter side of the input matrix."